Add design/20230601.gateway-route-hostnames. (#6124)
cert-manager/cert-manager: design/20230601.gateway-route-hostnames.md

A listener without hostname matches any hostname in attached routes, and they can simply be copied once for each route hostname.
I.e. the second example under [Proposal](#proposal) would be translated to the first.

Some glue data types are needed to support all routes that can carry hostnames.
At the moment, these are:

* [GRPCRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.GRPCRoute)
* [HTTPRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRoute)
* [TLSRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.TLSRoute)

### Test Plan

Since this code deals with how cert-manager reacts to changes in CRDs, it is enough to focus on unit tests.
For a given set of `Gateway`s and Routes, it should generate a given synthetic `Gateway`.

### Graduation Criteria

N/A

### Upgrade / Downgrade Strategy

Downgrading to a cert-manager that does not support looking up hostnames in Routes may lead to unavailability.

The change is upgrade-compatible, if all `Gateway`s already specify hostnames in `Listener`s.

> However, if some `Gateway` does not specify hostnames in `Listener`s, upgrading may cause certificates to be issued for hostnames not previously seen.

Reviewer: ```suggestion
However, if a `Gateway` does not specify hostnames in `Listener`s, upgrading may cause certificates to be issued for hostnames not previously seen.
```

In terms of security, this is not an issue, as the Routes have always existed; they simply didn't have a valid certificate.

### Supported Versions

* [Gateway API](https://gateway-api.sigs.k8s.io/references/spec/) is currently at v1beta1.

## Production Readiness

### How can this feature be enabled / disabled for an existing cert-manager installation?

Since this can be implemented as an input transform, that transform could be behind a feature flag.
Indeed, the entire Gateway API support is already behind a feature flag: `ExperimentalGatewayAPISupport`

It probably does not need a specific one.

### Does this feature depend on any specific services running in the cluster?

It requires Gateway API CRDs, but nothing more than is already required for gateway-shim.

> ### Will enabling / using this feature result in new API calls (i.e to Kubernetes apiserver or external services)?

Reviewer: ```suggestion
### Will enabling / using this feature result in new API calls (i.e. to Kubernetes apiserver or external services)?
```

> It will subscribe to \*Route resources, and not just `Gateway` resources.

Author: Sorry, how do you mean? The asterisk was meant to mean each of the HTTPRoute, TLSRoute, etc. resources.

```suggestion
It will subscribe to HTTPRoute, TLSRoute, and similar resources, and not just `Gateway` resources.
```

> It will subscribe to `HTTPRoute`, `TLSRoute`, and similar resources, and not just `Gateway` resources.

Reviewer: ```suggestion
It will subscribe to `HTTPRoute`, `TLSRoute`, and similar resources, beyond the previously subscribed `Gateway` resources.
```
